### DISCUSSION TITLE
Bug GA4 depuis le 31/05/2024 en rapport avec Consent Mode v2 ?

### DISCUSSION THREAD

**Discussion by Ludosane** [AmauriC/tarteaucitron.js#1254] [General]
Bonjour,
Je commence déjà par vous remercier de nous mettre à disposition votre CMP, ça rend vraiment bien service.
Depuis le 31/05 mon site (https://urlz.fr/qX5v) récolte très peu de visites dans Google Analytics 4 alors que je n'ai fait aucune modification depuis des mois. Après avoir cherché dans tous les sens (hébergeur, paramétrage GA) et même, désespéré, ajouté le code GA4 en dur (ce qui m'a ramené plus de visites mais toujours pas comme avant), je me demande si l'arrivée du Consent Mode V2 de Google n'en serait pas la cause ? Et obligerait peut-être à faire une mise à jour du script TAC ? Avez-vous des infos sur ce sujet ou une piste sur cet étrange bug ?
Merci d'avance
Ludovic

**Comment by AmauriC** [comment]
Bonjour,
En effet le consent mode impose une mise à jour du script pour un recueil de consentement conforme. 

tarteaucitron supporte le consent mode v2, il suffit de mettre à jour la librairie et de charger GA4 via tac comme avant.

**Comment by Ludosane** [comment]
Merci pour votre réponse, je vais de ce pas faire le changement.

**Comment by Ludosane** [comment]
J'ai juste encore une question au sujet de la libraire : je dois mettre à jour ces 2 fichiers ou juste un des 2 ?
- tarteaucitron.js
- tarteaucitron.services.js

**Reply by AmauriC** [reply]
Tous, le principal le .services, les langues et le css.

**Comment by Ludosane** [comment]
Remplacer la CSS est-ce vraiment obligatoire ? car je l'ai personnalisée... :-(

Et je vois que dans le 1er script, il y a aussi du changement avec cette ligne entre autres :
 "googleConsentMode": true, /* Enable Google Consent Mode v2 for Google ads and GA4 */

**Comment by AmauriC** [comment]
Modifier le fichier principal n'est pas une bonne chose.
Surtout pour du CSS, vous pouvez faire les modifications dans un fichier à part.

Vous pouvez ne pas le mettre à jour mais il peut y avoir des problèmes. 

L'initialisation du script comporte en effet régulièrement de nouvelles options.

**Comment by Ludosane** [comment]
Ce n'est pas une super nouvelle car me voilà reparti pour une refonte de la CMP mais en tout merci à vous de m'avoir répondu aussi vite.

**Comment by AmauriC** [comment]
tarteaucitron est la CMP. Si vous déployez une version modifié de tac sur vos sites il faut penser aux mises à jours et le prévoir : /

Le CSS peux sans souci être modifié mais dans un fichier à part.

C'est pareil pour toutes les librairies, il ne faut jamais modifier les fichiers sources.

**Comment by Ludosane** [comment]
Bonjour,
j'ai mis en ligne vos nouveaux fichiers, en abandonnant ma personnalisation, mais depuis je ne reçois plus de visites dans GA4. 
Savez-vous si on doit faire des modifications dans GA4 ? J'ai coché ce matin dans GA4 "vérifier les paramètres de consentement" et "confirmation de la collecte de données utilisateurs" mais j'ai toujours zéro visite en "Temps réel". Auriez-vous une idée de ce qui peut bloquer ?
Merci d'avance pour votre précieuse aide.
Ludovic

**Comment by Ludosane** [comment]
Je me demande aussi si mon problème ne vient pas du code que ma régie publicitaire m'a fait insérer pour gérer les bannières pubs et qui ressemble à ça :

<script async src="https://securepubads.g.doubleclick.net/tag/js/gpt.js"></script>
<script>
window.googletag = window.googletag || {cmd: []};
googletag.cmd.push(function() {
googletag.defineSlot('/228216569,205069399/regie-xxxx/Autres/Autres/Banniere-Haute', ['fluid'], 'div-gpt-ad-1695745409942-0').addService(googletag.pubads());
googletag.pubads().enableSingleRequest();
googletag.pubads().collapseEmptyDivs();
googletag.enableServices();
  });</script>
  
ça  a l'air d'être un code de Google Ad Manager et je ne vois pas comment le mettre dans votre CMP. Mais d'après vous est-ce que la présence de ce code pourrait bloquer Google Analytics ? Et ce qui est fou c'est que je ne trouve aucune explication claire sur le net sur les modifs à faire dans GA 4 à cause de cette satané Google Mode Consent v2 et j'y ai passé déjà pas mal d'heures. :-(

**Reply by AmauriC** [reply]
Je ne suis pas expert GA4 mais je ne pense pas qu'il y ai quelque chose à faire dans la console.
Sur le journaldesarts les signaux sont bien envoyées quand GA4 est activé via tarteaucitron.

**Reply by Ludosane** [reply]
Merci pour votre réponse rapide, ça fait toujours plaisir :-) Mais vous m'étonnez en me disant que les signaux sont bien envoyés, puis-je vous demander de me dire comment vous faites pour voir cela ? Personnellement j'utilise la console de Chrome et je ne vois pas la requête "collect?=v" propre à GA4 ni les .js et .css de tarteaucitron ; d'où mon étonnement. Mais d'après ce que vous me dîtes il y aurait du positif.

**Reply by AmauriC** [reply]
La variable `dataLayer` retourne bien le consent granted:
![Capture d’écran 2024-06-27 à 16 50 30](https://github.com/AmauriC/tarteaucitron.js/assets/8764074/b4373820-bd42-449c-81a7-7b397f7d9408)

GA4 est chargé:
![Capture d’écran 2024-06-27 à 16 49 54](https://github.com/AmauriC/tarteaucitron.js/assets/8764074/cf890893-7081-4f45-b1ee-ecaddf82e034)

Le CSS/JS de tac est regroupé dans un fichier global:
![Capture d’écran 2024-06-27 à 16 47 33](https://github.com/AmauriC/tarteaucitron.js/assets/8764074/73f19da9-06f3-4966-b6cd-e4817abbbce6)

Par contre en effet pas de `collect`.
Vous êtes sûr de l'identifiant GA4?

**Comment by Ludosane** [comment]
Malheureusement oui, je n'y comprend plus rien à moins que je n'utilise pas le bon service dans TAC ?
dans GA4 : G-H07J8JZQPJ
dans TAC : tarteaucitron.user.gtagUa = 'G-H07J8JZQPJ';

**Reply by AmauriC** [reply]
C'est le bon service.

**Reply by AmauriC** [reply]
En mettant le code GA4 directement sur le site sans passer par tarteaucitron ça fonctionne?

**Reply by Ludosane** [reply]
C'est ce que je fais l'autre jour pour voir si la CMP ne posait pas de problème et j'ai récupéré 50 % de mes visites habituelles.

**Reply by Ludosane** [reply]
Je viens de refaire un test en mettant le code GA4 en dur sur mon site de Dev et je vois la requête "Collect". Je l'enlève et je ne vois plus rien. J'ai pensé que l'ordre des appels .js avait peut-être une importance (étant sur Drupal, je les mets dans un fichier "xxxx.libraries.yml") alors j'ai testé en mettant chaque .min au dessus de son pendant mais cela n'a rien changé. Bref j'ai essayé tellement de trucs depuis quelques jours que là je vais aller me boire un petit verre de rosé :-) S'il vous vient une idée qui pourrait m'aider, elle sera la plus que bienvenue. Demain je vérifierai tous vos fichiers car j'en ai modifié quelques-uns et c'est là peut-être le souci ? Bonne soirée en tout cas.
Ludovic

**Reply by AmauriC** [reply]
C'est vraiment étrange. Tenez moi au courant, je regarderai de nouveau demain.

Bon apéro, pour moi aussi 😎

**Reply by Ludosane** [reply]
J'espère que l'apéro fut bon, j'ai bien apprécié le mien hier. 😎

Cet après-midi, plein d'espoir, j'ai transféré vos fichiers .js un par un sur mon Dev pour voir si la requête "Collect" se lançait et malheureusement non.
J'ai tenté ensuite, même si je  n'y croyais pas, de transférer les fichiers et dossier que je n'avais pas utilisé (github, gitignore, license, package.json, Readme et même Security.md) mais toujours sans résultat. Et j'avoue que là je commence à être en manque d'idée :-(

**Reply by AmauriC** [reply]
Je sèche 🧐 Essai de contacter l'assistance Google, il y a sûrement un problème de réglage sur le compte. 

L'intégration côté tarteaucitron est la même que sur tous les autres sites et je n'ai pas d'autres retours similaires..

**Reply by Ludosane** [reply]
Contacter l'assistance de Google Analytics... alors là je suis dead :-( J'ai déjà essayé et n'ai eu absolument aucune réponse sur leur forum. Je vais essayer l'appel à un ami. Et s'il trouve la solution, je reviendrais l'écrire ici. Merci en tout cas pour ton suivi et re-bon apéro car l'heure approche 😎

**Reply by Ludosane** [reply]
Bonjour Amauric,
L'appel à un ami a marché ! Cet excellent développeur a trouvé 2 choses. La 1ère c'est que j'appelais 2x tarteaucitron (le normal et le min.js) mais surtout j'avais un problème beaucoup plus obscure de surcharge JavaScript due à une fonction créée en 2016 par une webagency qui avait appelé sa fonction "Map" et depuis la fonction "Map" est devenu un standard chez Javascript. Il a donc modifié le nom de cette fonction et tout est rentré dans l'ordre ! ça je ne l'aurai jamais trouvé ! 
Sur cet excellente nouvelle je te souhaite un tout aussi excellent WE !
Ludovic

**Reply by AmauriC** [reply]
Génial! Merci pour le retour. Bon weekend! 😎
